src/rest/rest_web_server.cpp

```diff
@@ -0,0 +1,229 @@
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "rest/rest_web_server.hpp"
+
+#include <cerrno>
+
+#include <fcntl.h>
+
+using std::chrono::duration_cast;
+using std::chrono::microseconds;
+using std::chrono::steady_clock;
+
+namespace otbr {
+namespace rest {
+
+// Maximum number of connection a server support at the same time.
+static const uint32_t kMaxServeNum = 500;
+// Port number used by Rest server.
+static const uint32_t kPortNumber = 8081;
+
+std::unique_ptr<RestWebServer> RestWebServer::sRestWebServer;
+
+RestWebServer::RestWebServer(ControllerOpenThread *aNcp)
+    : mResource(Resource(aNcp))
+{
+}
+
+RestWebServer *RestWebServer::GetRestWebServer(ControllerOpenThread *aNcp)
+{
+    if (sRestWebServer.get() == nullptr)
+    {
+        sRestWebServer = std::unique_ptr<RestWebServer>(new RestWebServer(aNcp));
+    }
+    return sRestWebServer.get();
+}
+
+otbrError RestWebServer::Init(void)
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    mResource.Init();
+
+    error = InitializeListenFd();
+
+    return error;
+}
+
+void RestWebServer::UpdateFdSet(otSysMainloopContext &aMainloop)
+{
+    FD_SET(mListenFd, &aMainloop.mReadFdSet);
+    aMainloop.mMaxFd = aMainloop.mMaxFd < mListenFd ? mListenFd : aMainloop.mMaxFd;
+
+    for (auto it = mConnectionSet.begin(); it != mConnectionSet.end(); ++it)
+    {
+        Connection *connection = it->second.get();
+        connection->UpdateFdSet(aMainloop);
+    }
+}
+
+otbrError RestWebServer::Process(otSysMainloopContext &aMainloop)
+{
+    otbrError   error = OTBR_ERROR_NONE;
+    Connection *connection;
+
+    error = UpdateConnections(aMainloop.mReadFdSet);
+
+    for (auto it = mConnectionSet.begin(); it != mConnectionSet.end(); ++it)
+    {
+        connection = it->second.get();
+        error      = connection->Process(aMainloop.mReadFdSet, aMainloop.mWriteFdSet);
+    }
+
+    return error;
+}
+
+otbrError RestWebServer::UpdateConnections(fd_set &aReadFdSet)
+{
+    otbrError error   = OTBR_ERROR_NONE;
+    auto      eraseIt = mConnectionSet.begin();
+
+    // Erase useless connections
+    for (eraseIt = mConnectionSet.begin(); eraseIt != mConnectionSet.end();)
+    {
+        Connection *connection = eraseIt->second.get();
+
+        if (connection->IsComplete())
+        {
+            eraseIt = mConnectionSet.erase(eraseIt);
+        }
+        else
+        {
+            eraseIt++;
+        }
+    }
+
+    // Create new connection if listenfd is set
```

Comment:
When  calling `Accept` in `UpdateConnections`( it is also the last step of `UpdateConnections`), it will first check the retval(fd) of system call `accept`,  then try to create new connection, if success, the action that following `UpdateConnections` is to process each connection including the one just created. if failed, it will close it, seems I forgot to close it if `setnonblocking` fail, will fix it.